src/service.rs

```diff
@@ -0,0 +1,327 @@
+};
+use crate::xds::istio::workload::PortList;
+use bytes::Bytes;
+use std::collections::{HashMap, HashSet};
+use std::sync::{Arc, RwLock};
+use tracing::trace;
+
+use xds::istio::workload::Service as XdsService;
+
+use crate::xds;
+
+#[derive(Debug, Eq, PartialEq, Clone, serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub struct Service {
+    pub name: String,
+    pub namespace: String,
+    pub hostname: String,
+    pub addresses: Vec<NetworkAddress>,
+    pub ports: HashMap<u16, u16>,
+    #[serde(default)]
+    pub endpoints: HashMap<NetworkAddress, Endpoint>,
+}
+
+impl Service {
+    pub fn namespaced_hostname(&self) -> NamespacedHostname {
+        NamespacedHostname {
+            namespace: self.namespace.clone(),
+            hostname: self.hostname.clone(),
+        }
+    }
+}
+
+#[derive(Debug, Eq, PartialEq, Clone, serde::Serialize, serde::Deserialize)]
+#[serde(rename_all = "camelCase", deny_unknown_fields)]
+pub struct Endpoint {
+    pub address: NetworkAddress,
+    pub port: HashMap<u16, u16>,
+}
+
+impl TryFrom<&XdsService> for Service {
+    type Error = WorkloadError;
+
+    fn try_from(s: &XdsService) -> Result<Self, Self::Error> {
+        let mut nw_addrs = Vec::new();
+        for addr in &s.addresses {
+            let network_address = network_addr(
+                &addr.network,
+                byte_to_ip(&Bytes::copy_from_slice(&addr.address))?,
+            );
+            nw_addrs.push(network_address);
+        }
+        let svc = Service {
+            name: s.name.to_string(),
+            namespace: s.namespace.to_string(),
+            hostname: s.hostname.to_string(),
+            addresses: nw_addrs,
+            ports: (&PortList {
+                ports: s.ports.clone(),
+            })
+                .into(),
+            endpoints: Default::default(), // intentionally empty; will be populated once inserted into the workload store
+        };
+        Ok(svc)
+    }
+}
+
+/// Data store for service information.
+#[derive(serde::Serialize, Default, Debug)]
+pub struct ServiceStore {
+    /// Maintains a mapping of service IP -> (workload IP -> workload endpoint)
+    /// this is used to handle ordering issues if workloads with VIPs are received before services.
+    staged_vips: HashMap<NetworkAddress, HashMap<NetworkAddress, Endpoint>>,
+
+    /// Maintains a mapping of workload IP to VIP. This is used only to handle removal of service
+    /// endpoints when a workload is removed.
+    workload_to_vips: HashMap<NetworkAddress, HashSet<NetworkAddress>>,
+
+    /// Allows for lookup of services by network address, the service's xds secondary key.
+    /// this map stores the same services as `services_by_hostname` so we can mutate the services in both maps
+    /// at once when a service is updated or workload updates affect a service's endpoints.
+    by_vip: HashMap<NetworkAddress, Arc<RwLock<Service>>>,
```

Comment:
next pr should remove the `Arc<RwLock` :slightly_smiling_face:

Reply:
Yeah, I'm not a fan of that either ... any idea why this was done?

Reply:
Something about feeling more confident in keeping the two maps in sync if they share the same object, but I think it can be done without

Reply:
it was done for these three cases (which are covered by unit tests)
- https://github.com/istio/ztunnel/blob/b64f8255cc7baa64549384e083c0965114fbd177/src/workload.rs#L1443-L1445
- https://github.com/istio/ztunnel/blob/b64f8255cc7baa64549384e083c0965114fbd177/src/workload.rs#L1504-L1506
- https://github.com/istio/ztunnel/blob/b64f8255cc7baa64549384e083c0965114fbd177/src/workload.rs#L1531-L1533

either we need to perform this mutation on every copy of the stored data per index, or put the same data behind the index (i.e. by network/IP and ns/hostname). this opted to put the same data behind the index without pulling in new deps

perhaps a cleaner way of doing this would be something like https://docs.rs/double-map/latest/double_map/